opensrp-app/src/main/java/org/smartregister/util/FormUtils.java

```diff
@@ -1130,11 +1135,64 @@ private void createNewClientDocument(org.smartregister.cloudant.models.Client cl
         mCloudantDataHandler.createClientDocument(client);
     }
 
-    public JSONObject getFormJsonFromRepositoryOrAssets(String formIdentity) {
-        ClientFormRepository clientFormRepository = CoreLibrary.getInstance().context().getClientFormRepository();
-        String locale = mContext.getResources().getConfiguration().locale.getLanguage();
+    @Nullable
+    public JSONObject getFormJsonFromRepositoryOrAssets(@NonNull String formIdentity) {
+        return getFormJsonFromRepositoryOrAssetsWithOptionalCallback(formIdentity, null);
+    }
+
+    public void getFormJsonFromRepositoryOrAssets(@NonNull String formIdentity, @Nullable OnFormFetchedCallback<JSONObject> onFormFetchedCallback) {
+        getFormJsonFromRepositoryOrAssetsWithOptionalCallback(formIdentity, onFormFetchedCallback);
+    }
+
+    private JSONObject getFormJsonFromRepositoryOrAssetsWithOptionalCallback(String formIdentity, @Nullable OnFormFetchedCallback<JSONObject> onFormFetchedCallback) {
+        ClientForm clientForm = getClientFormFromRepository(formIdentity);
+
+        try {
+            if (clientForm != null) {
+                Timber.d("============%s form loaded from db============", formIdentity);
+
+                JSONObject formJson = new JSONObject(clientForm.getJson());
+                injectFormStatus(formJson, clientForm);
+
+                if (onFormFetchedCallback != null) {
+                    onFormFetchedCallback.onFormFetched(formJson);
+                    return null;
+                }  else {
+                    return formJson;
+                }
+            }
+        } catch (JSONException e) {
+            Timber.e(e);
+
+            if (onFormFetchedCallback != null) {
+                handleJsonFormOrRulesError(false, formIdentity, form -> {
+                    try {
+                        JSONObject jsonObject = form == null ? null : new JSONObject(form);
+                        onFormFetchedCallback.onFormFetched(jsonObject);
+                    } catch (JSONException ex) {
+                        Timber.e(ex);
+                    }
+                });
+            } else {
+                return null;
+            }
+        }
+
+        Timber.d("============%s form loaded from Assets=============", formIdentity);
+        JSONObject jsonObject = getFormJson(formIdentity);
 
+        if (onFormFetchedCallback != null) {
+            onFormFetchedCallback.onFormFetched(jsonObject);
+            return null;
+        }  else {
+            return jsonObject;
+        }
+    }
+
+    private ClientForm getClientFormFromRepository(String formIdentity) {
```

Comment:
Does this account for MLS v2.0 where even if the locale changes we still the same form?

Reply:
Nope, I probably didn't get this question. Can you table this for the meeting? 

I have not put a lot of effort on MLS v2 support due to the inability to clearly test but as long as the property files are considered as forms, the idea is that it would only require enabling the DB as a source of property files also(instead of just the assets)

Reply:
Added, thanks.